README.txt
Python interface to split music sets into individual tracks using tracklist link from 1001 tracklists
    - geckodriver, ffmpeg, and fpcalc must be installed and added to path

changelog.txt
v01: Initial Commit

v01.1:
    set_split.py:
        - modified back end to terminate working thread at end of splitting process to allow for re-runs with new fields
        - functionality for offsetting timings according to GUI input implemented

        .format_tracklist:
            - comments added for clarity
            - filtering exception added for lines found in tracklists with "Artist on stage"

v02:
    - handling of tracklists overhauled
        - copy and pasting of exported tracklist from 1001 tracklists no longer required, just website link
    - UI streamlined to handle overhaul
    - new yml exported
    - README updated

    set_split.py:
        - functionality for tracklist importing from 1001 tracklists link added
            - uses similar threading protocol as previously implemented in main method
        - variable renaming
        - spacing made uniform

   layout.py:
        - elements added to handle new tracklist importing functionality
        - layout modified to be better suited to new tracklisting functionality
        - visual elements grouped in code in initialization
        - spacing made uniform

   .gitignore:
        - update to ignore geckodriver logs

v02.1: 
    musiC.atalogue.yml:
        - update yml file to be compatible across all operating systems

    set_split.py:
        - change tracklist handling to successfully remove label name in edge case of mashup
        - missing timing tags will now be appended to previous track rather than crashing the process

v03:
    set_split.py:
        - mp3 file handling changed from pydub to ffmpeg allowing for splitting of larger files (FFMPEG MUST BE IN PATH FOR PROGRAM TO WORK)
        - deprecated functions labeled as such, all pydub handling removed
        - importing changed to more intuitive order

    musiC.atalogue.yml:
        - updated dependencies to reflect new back end 2handling

    layout.py:
        - offsetting will be converted to seconds, changed to reflect as such in interface
        - added window title

v03.1:
    set_split.py:
        - argument numbers across functions fixed to be consistent with new file handling
        - filtering exception added for lines found in tracklists with "set"
        - debug tracklist handling issues


v03.1.1:
    musiC.atalogue.yml:
        - update dependencies to allow environment to be compatible with zspotify

v04 (in progress):
    complete overhaul, turning set_split into command line interface
    - GUI is now deprecated and added to directory /deprecated_gUI

    musiC.a.yml:
        - yml file renamed and dependencies updated

    media_editing.py:
        - functions for downloading and splitting set according to arguments given

    trasklist_editing.py:
        - functions for pulling tracklist from 1001 tracklists and splitting them into tags for splitting

    set_split.py:
        - command line wrapper for set_split

    test_fingerprint.py
        - BEGINNING OF/TESTING OF FINGERPRINT RECOGNITION.

    README.txt
        - specify FFMPEG and fpcalc requirements


TODO:
    - REGEX for tracklist handling
    - Add documentation to README

test_fingerprinting.py
#%%

file_name = "E:/Library v4/Trance/Paul van Dyk/The Politics Of Dancing 3/10 - For You.mp3"
file_name_2 ='E:/Library v4/Other/Scores/Risk of Rain 2/15 - The Rain Formerly Known as Purple.mp3'

#%%
import acoustid
import requests
import json

fingerprint = acoustid.fingerprint_file(file_name_2)
response = requests.get("https://api.acoustid.org/v2/lookup", params={"client": "dASxAPhfZl", "meta": "recordings", "duration": int(fingerprint[0]), "fingerprint": fingerprint[1], "format": "json",})
data = response.json()

set_split.py
import argparse
import types

import media_editing
import tracklist_editing


def arg_parser():

    parser = argparse.ArgumentParser(description = 'Downloader')
    parser.add_argument('-i', '--input_directory', type = str, help = 'directory to input file')
    parser.add_argument('-l', '--media_link', type = str, help = 'media link to be downloaded')
    parser.add_argument('-o', '--output_directory', type=str, help='directory to save files if downloading link')
    parser.add_argument('-f', '--file_name', type=str, help='name to be given to file, if downloading')
    parser.add_argument('-n', '--set_title', type = str, help = 'title to tag')
    parser.add_argument('-aa', '--album_artist', type = str, help = 'album artist to tag')
    parser.add_argument('-a', '--album', type = str, help = 'album to tag')
    parser.add_argument('-g', '--genre', type = str, help = 'genre to tag')
    parser.add_argument('-t', '--tracklist_link', type=str, help='tracklist link, to setlist to split set if given')
    parser.add_argument('-d', '--offset', type = float, help = 'offset in seconds to add (+) or subtract(-) from each timing after the first')
    args = parser.parse_args()

    return args

# def create_dummy_args():
#
#     args = types.SimpleNamespace()
#     args.media_link = "https://www.youtube.com/watch?v=haWujR4wurw"
#     args.tracklist_link = "https://www.1001tracklists.com/tracklist/7ml9bht/tiesto-mainstage-ultra-music-festival-miami-united-states-2015-03-27.html"
#     args.output_directory = "/home/c/Downloads/"
#     args.genre = 'Big Room'
#     args.album_artist = 'Tiesto'
#     args.album = "Ultra 2015"
#     args.file_name = 'Tiest Ultra 2015.mp3'
#
#     return args



if __name__ == "__main__":

    args = arg_parser()

    if args.media_link:
        full_file_name = media_editing.download_media(args.media_link, args.output_directory, args.file_name, args)

    if args.tracklist_link:

        if args.offset:
            offset = args.offset
        else:
            offset = 0

        fail = True
        while fail:
            try:
                track_list = tracklist_editing.get_tracklist(args.tracklist_link)
                fail = False
            except:
                pass

        [timings, titles, artists] = tracklist_editing.format_tracklist(track_list, offset = offset)

        if args.input_directory:
            media_editing.split_set(args.input_directory, timings, titles, artists, args)
        else:
            media_editing.split_set(full_file_name, timings, titles, artists, args)

media_editing.py
import os
import subprocess
from pathlib import Path
import shutil

def download_media(link, output_directory, file_name, tag_object):

    [set_title, album_artist, album, genre] = tag_parser(tag_object)

    download_file_name = output_directory + 'temp' + file_name
    full_file_name = output_directory + file_name
    subprocess.call(['yt-dlp', '--newline', '-i' , '-o', download_file_name, '-x', '--audio-format', 'mp3', '--ignore-config', '--hls-prefer-native', link])
    subprocess.call(['ffmpeg' , '-i',  download_file_name, '-map' , '0' , '-y', '-write_id3v2', '1',
                     '-metadata', 'title='+set_title, '-metadata', 'album_artist='+album_artist, '-metadata', 'album='+album
                     , '-metadata', 'genre='+genre, full_file_name])
    os.remove(download_file_name)

    return full_file_name



def split_set(input_file, timings, titles, artists, tag_object):

    [set_title, album_artist, album, genre] = tag_parser(tag_object)

    input_file_parent =  str(Path(input_file).parent.absolute())+'/'
    input_file_base = os.path.basename(input_file)

    if album != '':
        output_path_parent = input_file_parent + album + '/'
    else:
        output_path_parent = input_file[:-4] + '/'

    os.makedirs(output_path_parent, exist_ok = True)
    shutil.copyfile(input_file, output_path_parent + input_file_base)

    output_path = output_path_parent + '/split/'
    os.makedirs(output_path, exist_ok = True)


    for i in range(len(timings)):

        if i == len(timings) - 1:
            subprocess.call(
                ['ffmpeg', '-y', '-ss', timings[i], '-i', input_file,
                 '-metadata', 'title=' + titles[i], '-metadata', 'album_artist=' + album_artist,
                 '-metadata', 'album=' + album, '-metadata', 'artist='+artists[i], '-metadata',
                 'track=' + str(i + 1), '-metadata', 'genre=' + genre,'-c', 'copy',
                 output_path + str(i + 1) + ' - ' + titles[i] + '.mp3'])

        else:

            subprocess.call(['ffmpeg', '-y', '-ss', timings[i], '-to', timings[i+1], '-i', input_file,
                            '-metadata', 'title='+titles[i], '-metadata', 'album_artist='+album_artist,
                             '-metadata', 'album='+album, '-metadata', 'artist='+artists[i],
                             '-metadata', 'track='+str(i+1), '-metadata', 'genre='+genre, '-c', 'copy',
                             output_path + str(i + 1) + ' - ' + titles[i] + '.mp3'])

    os.remove(input_file)



def tag_parser(object):

    if object.set_title:
        set_title = object.set_title
    else:
        set_title = object.file_name[:-4]

    if object.album_artist:
        album_artist = object.album_artist
    else:
        album_artist = ''

    if object.album:
        album = object.album
    else:
        album = ''

    if object.genre:
        genre = object.genre
    else:
        genre = ''

    return set_title, album_artist, album, genre

tracklist_editing.py
from os import mkdir,remove

import datetime
import time

from selenium import webdriver
from selenium.webdriver.common.by import By

def get_tracklist(link):
    email_1001 = 'dcarbonator@gmail.com'
    password_1001 = 'musicislife'
    driver = webdriver.Firefox()
    driver.set_window_size(1080, 1920)
    driver.get(link)
    driver.find_element(By.CSS_SELECTOR, ".fa-sign-in").click()
    driver.find_element_by_name('email').send_keys(email_1001)
    driver.find_element_by_name('password').send_keys(password_1001)
    driver.find_element_by_name('login').click()
    time.sleep(2.5)
    driver.find_element_by_css_selector("[title*='export tracklist']").click()
    time.sleep(2.5)
    track_list = driver.find_element(By.ID, 'msgBoxTxt').text
    driver.find_element_by_name('btn_msgpane_ok').click()
    driver.quit()

    return track_list



def format_tracklist(track_list, offset = 0):  # function to format tracklist for tags

    text = []

    with open('tracklist.txt', 'wb') as file:
        file.write(track_list.encode('utf-8'))

    with open('tracklist.txt') as file:
        for line in file:
            text.append(line)

    remove('tracklist.txt')

    text = text[2:]
    text = text[:-2]

    timings = []
    titles = []
    artists = []
    title_string = []
    artist_string = []

    for i in range(len(text)):

        line = text[i]

        if line[0] == '[':  # check for line starting with timestamp

            if line[1] != '?':  # check for missing timing tag

                timings.append(line[1:line.find(']')])  # grab song timing from inside brackets
                line = line[line.find(']') + 1:]  # remove timestamp before first closing bracket

                if len(title_string) > 128:
                    title_string = title_string[0:128]

                # append previously formatted tag before writing new one, used to have full string, if multiple songs or artists are in same line
                titles.append(title_string)
                artists.append(artist_string)

                title_string = remove_track_label(line)
                artist_string = remove_edge_spaces(line[0:line.rfind(' - ')])

            else:
                line = line[line.find(']') + 1:]  # remove faulty timestamp before first closing bracket
                title_string = title_string + ' & ' + remove_track_label(line)
                artist_string = artist_string + '; ' + remove_edge_spaces(str(line[0:line.find(' - ')]))

        elif "on stage" not in str.lower(line) and "set" not in str.lower(
                line):  # add filter for pasted lines that only have artists as "on stage" and different "sets" in a singular set

            title_string = title_string + ' & ' + remove_track_label(line)
            artist_string = artist_string + '; ' + remove_edge_spaces(str(line[3:line.find(' - ')]))

    titles.append(title_string)
    artists.append(artist_string)

    timings = offset_timings(timings, offset)

    # first entries are empty, remove them
    del titles[0]
    del artists[0]

    return timings, titles, artists



def remove_edge_spaces(string):
    if string[0] == ' ':
        string = string[1:]
    if string[-1] == ' ':
        string = string[:-1]
    return string



def offset_timings(timings, offset):
    seconds_timings = []
    for i in range(len(timings)):
        if len(timings[i].split(':')) == 1:
            seconds_timings.append(timings[i])
        elif len(timings[i].split(':')) == 2:
            m, s = timings[i].split(':')
            seconds_timings.append(int(m) * 60 + int(s))
        elif len(timings[i].split(':')) == 3:
            h, m, s = timings[i].split(':')
            seconds_timings.append(int(h) * 3600 + int(m) * 60 + int(s))
        else:
            print('ERROR')

    seconds_offset_timings = [timing + offset for timing in seconds_timings]
    timings_offset = []

    for timing in seconds_offset_timings:
        timings_offset.append(datetime.timedelta(seconds=timing))

    timings_offset[0] = '0:00'

    timings_offset = [str(timing) for timing in timings_offset]

    return timings_offset



def remove_track_label(string):
    if string.find('Mashup)') != -1:
        string = remove_edge_spaces(string[string.find(' - ') + 2:string.find('Mashup)') + 7])
    else:
        string = remove_edge_spaces(string[string.find(' - ') + 2:string.find('[')])

    return string

layout.py
# set_split Front End Initializer
# Created by: PyQt5 UI code generator 5.15.2
# C.

from PyQt5 import QtCore, QtGui, QtWidgets



class Ui_Dialog(object):

    def setupUi(self, Dialog):

        Dialog.setObjectName("Dialog")
        Dialog.resize(884, 951)

        self.title = QtWidgets.QLabel(Dialog)
        self.title.setGeometry(QtCore.QRect(10, 10, 201, 61))
        font = QtGui.QFont()
        font.setPointSize(32)
        font.setBold(True)
        font.setItalic(False)
        font.setWeight(75)
        self.title.setFont(font)
        self.title.setObjectName("title")

        self.tracklist_label = QtWidgets.QLabel(Dialog)
        self.tracklist_label.setGeometry(QtCore.QRect(440, 20, 111, 31))
        font = QtGui.QFont()
        font.setPointSize(20)
        self.tracklist_label.setFont(font)
        self.tracklist_label.setObjectName("tracklist_label")

        self.tracklist_link = QtWidgets.QTextEdit(Dialog)
        self.tracklist_link.setGeometry(QtCore.QRect(440, 55, 300, 35))
        font = QtGui.QFont()
        font.setPointSize(10)
        self.tracklist_link.setFont(font)
        self.tracklist_link.setObjectName("tracklist_link")

        self.tracklist_update = QtWidgets.QPushButton(Dialog)
        self.tracklist_update.setGeometry(QtCore.QRect(760, 55, 100, 35))
        font = QtGui.QFont()
        font.setPointSize(16)
        self.tracklist_update.setFont(font)
        self.tracklist_update.setObjectName("update")

        self.tracklist_disp = QtWidgets.QTextBrowser(Dialog)
        self.tracklist_disp.setGeometry(QtCore.QRect(440, 100, 421, 631))
        font = QtGui.QFont()
        font.setPointSize(12)
        self.tracklist_disp.setFont(font)
        self.tracklist_disp.setObjectName("tracklist_disp")

        self.get_file = QtWidgets.QPushButton(Dialog)
        self.get_file.setGeometry(QtCore.QRect(15, 100, 151, 71))
        font = QtGui.QFont()
        font.setPointSize(16)
        self.get_file.setFont(font)
        self.get_file.setObjectName("get_file")

        self.file_path = QtWidgets.QLabel(Dialog)
        self.file_path.setGeometry(QtCore.QRect(15, 175, 384, 61))
        font = QtGui.QFont()
        font.setPointSize(12)
        self.file_path.setFont(font)
        self.file_path.setObjectName("file_path")

        self.output_label = QtWidgets.QLabel(Dialog)
        self.output_label.setGeometry(QtCore.QRect(12, 250, 256, 61))
        font = QtGui.QFont()
        font.setPointSize(20)
        self.output_label.setFont(font)
        self.output_label.setObjectName("output_label")

        self.output_dir = QtWidgets.QLabel(Dialog)
        self.output_dir.setGeometry(QtCore.QRect(15, 300, 384, 61))
        font = QtGui.QFont()
        font.setPointSize(12)
        self.output_dir.setFont(font)
        self.output_dir.setObjectName("output_dir")

        self.album_artist_input = QtWidgets.QTextEdit(Dialog)
        self.album_artist_input.setGeometry(QtCore.QRect(440, 790, 211, 41))
        font = QtGui.QFont()
        font.setPointSize(18)
        self.album_artist_input.setFont(font)
        self.album_artist_input.setObjectName("album_artist_input")

        self.album_artist_label = QtWidgets.QLabel(Dialog)
        self.album_artist_label.setGeometry(QtCore.QRect(440, 760, 161, 31))
        font = QtGui.QFont()
        font.setPointSize(20)
        self.album_artist_label.setFont(font)
        self.album_artist_label.setObjectName("album_artist_label")

        self.album_label = QtWidgets.QLabel(Dialog)
        self.album_label.setGeometry(QtCore.QRect(440, 850, 160, 31))
        font = QtGui.QFont()
        font.setPointSize(20)
        self.album_label.setFont(font)
        self.album_label.setObjectName("album_label")

        self.album_input = QtWidgets.QTextEdit(Dialog)
        self.album_input.setGeometry(QtCore.QRect(440, 880, 211, 41))
        font = QtGui.QFont()
        font.setPointSize(16)
        self.album_input.setFont(font)
        self.album_input.setObjectName("album_input")

        self.genre_label = QtWidgets.QLabel(Dialog)
        self.genre_label.setGeometry(QtCore.QRect(690, 850, 141, 31))
        font = QtGui.QFont()
        font.setPointSize(18)
        self.genre_label.setFont(font)
        self.genre_label.setObjectName("genre_label")

        self.genre_input = QtWidgets.QTextEdit(Dialog)
        self.genre_input.setGeometry(QtCore.QRect(690, 880, 151, 41))
        font = QtGui.QFont()
        font.setPointSize(18)
        self.genre_input.setFont(font)
        self.genre_input.setObjectName("genre_input")

        self.offset_label = QtWidgets.QLabel(Dialog)
        self.offset_label.setGeometry(QtCore.QRect(690, 760, 141, 31))
        font = QtGui.QFont()
        font.setPointSize(20)
        self.offset_label.setFont(font)
        self.offset_label.setObjectName("offset_label")

        self.offset_input = QtWidgets.QTextEdit(Dialog)
        self.offset_input.setGeometry(QtCore.QRect(690, 790, 151, 41))
        font = QtGui.QFont()
        font.setPointSize(18)
        self.offset_input.setFont(font)
        self.offset_input.setObjectName("offset_input")

        self.output_update = QtWidgets.QTextBrowser(Dialog)
        self.output_update.setGeometry(QtCore.QRect(40, 375, 380, 384))
        font = QtGui.QFont()
        font.setPointSize(12)
        self.output_update.setFont(font)
        self.output_update.setObjectName("output_update")

        self.run = QtWidgets.QPushButton(Dialog)
        self.run.setGeometry(QtCore.QRect(80, 780, 281, 141))
        font = QtGui.QFont()
        font.setPointSize(72)
        font.setBold(True)
        font.setWeight(75)
        self.run.setFont(font)
        self.run.setObjectName("run")

        self.retranslateUi(Dialog)
        QtCore.QMetaObject.connectSlotsByName(Dialog)



    def retranslateUi(self, Dialog):

        _translate = QtCore.QCoreApplication.translate
        Dialog.setWindowTitle(_translate("Dialog", "Dialog"))
        self.title.setText(_translate("Dialog", "Set Split"))
        self.get_file.setText(_translate("Dialog", "Select File"))
        self.output_label.setText(_translate("Dialog", "Output Directory:"))
        self.tracklist_label.setText(_translate("Dialog", "Tracklist:"))
        self.tracklist_disp.setText(_translate("Dialog", "Tracklist Will Show Here"))
        self.tracklist_update.setText(_translate("Dialog", "Update"))
        self.tracklist_link.setText(_translate("Dialog", "Paste 1001 tracklists Set Link Here"))
        self.album_artist_label.setText(_translate("Dialog", "Album Artist:"))
        self.album_label.setText(_translate("Dialog", "Album:"))
        self.genre_label.setText(_translate("Dialog", "Genre:"))
        self.offset_label.setText(_translate("Dialog", "Offset (s)"))
        self.run.setText(_translate("Dialog", "Run"))
        Dialog.setWindowTitle('Set Split')
